File: Python/103_Binary_Tree_Zigzag_Level_Order_Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if root == None:
            return []
        dd = {}
        self.preOrderTraversal(root, 0, dd)
        ll = []
        for i in dd:
            ll += [dd[i]]
        a = False
        for i in ll:
            if a:
                i.reverse()
            a = not a
        return ll
        
    def preOrderTraversal(self, root, l, d):
        if root != None:
            if l not in d:
                d[l] = [root.val]
            else:
                d[l].append(root.val)
            if root.left != None:
                self.preOrderTraversal(root.left, l + 1, d)
            if root.right != None:
                self.preOrderTraversal(root.right, l + 1, d)
        return d

File: C++/103_Binary_Tree_Zigzag_Level_Order_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(root == NULL) return vector<vector<int>>();
        unordered_map<int, vector<int>> d;
        preOrderTraversal(root, 0, d);
        int n = d.size();
        vector<vector<int>> ll(n);
        for(int i = 0; i < n; i++){
            ll[i] = d[i];
        }
        bool a = 0;
        for(auto &x : ll){
            if(a){
                reverse(x.begin(), x.end());
            }
            a = !a;
            cout << a << endl;
        }
        return ll;
    }
    void preOrderTraversal(TreeNode* root, int i, unordered_map<int, vector<int>>& d){
        if(root != NULL){
            d[i].push_back(root->val);
            if(root->left != NULL) preOrderTraversal(root->left, i+1, d);
            if(root->right != NULL) preOrderTraversal(root->right, i+1, d);
        }
    }
    
};